### ISSUE TITLE
btrfs vs ext4 PVC provisioning delay

### ISSUE THREAD

**Issue by jeanfabrice** [SynologyOpenSource/synology-csi#25]
Hi team, 

Using these StorageClass definitions:
```
---
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: test-btrfs
provisioner: csi.san.synology.com
parameters:
  location: '/volume1' # kubernetes SSD volume
  fsType: 'btrfs'
  thin_provisioning: "true"
reclaimPolicy: Delete
allowVolumeExpansion: true
```

and 
```
---
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: test-ext4
provisioner: csi.san.synology.com
parameters:
  location: '/volume1' # kubernetes SSD volume
  fsType: 'ext4'
  thin_provisioning: "true"
reclaimPolicy: Delete
allowVolumeExpansion: true
```

It takes nearly 2m for a `pod-ext4` Debian pod to start with a 50Gb ext4 PVC:
```
 Warning  FailedScheduling        2m12s  default-scheduler        0/6 nodes are available: 6 pod has unbound immediate PersistentVolumeClaims.
  Warning  FailedScheduling        2m10s  default-scheduler        0/6 nodes are available: 6 pod has unbound immediate PersistentVolumeClaims.
  Normal   Scheduled               2m8s   default-scheduler        Successfully assigned synology-csi/pod-ext4 to worker-2
  Normal   SuccessfulAttachVolume  2m7s   attachdetach-controller  AttachVolume.Attach succeeded for volume "pvc-67cbad0d-e431-4c7d-9342-b39e0cc175b4"
  Normal   Pulling                 18s    kubelet                  Pulling image "debian"
  Normal   Pulled                  17s    kubelet                  Successfully pulled image "debian" in 1.030883146s
  Normal   Created                 16s    kubelet                  Created container pod-ext4
  Normal   Started                 16s    kubelet                  Started container pod-ext4
```
<img width="1228" alt="image" src="https://user-images.githubusercontent.com/2949987/152677547-3cbe7dae-3c55-47e0-9049-bfd103f919b4.png">

While it takes only 12s for a `pod-btrfs` Debian pod to start with a 50Gb btrfs PVC:
```
  Type     Reason                  Age   From                     Message
  ----     ------                  ----  ----                     -------
  Warning  FailedScheduling        23s   default-scheduler        0/6 nodes are available: 6 pod has unbound immediate PersistentVolumeClaims.
  Warning  FailedScheduling        21s   default-scheduler        0/6 nodes are available: 6 pod has unbound immediate PersistentVolumeClaims.
  Normal   Scheduled               19s   default-scheduler        Successfully assigned synology-csi/pod-btrfs to worker-2
  Normal   SuccessfulAttachVolume  18s   attachdetach-controller  AttachVolume.Attach succeeded for volume "pvc-a4da0c08-ee82-4ae6-ad5c-feb3cf823ed7"
  Normal   Pulling                 6s    kubelet                  Pulling image "debian"
  Normal   Pulled                  5s    kubelet                  Successfully pulled image "debian" in 1.057510873s
  Normal   Created                 5s    kubelet                  Created container pod-btrfs
  Normal   Started                 5s    kubelet                  Started container pod-btrfs
```
<img width="1228" alt="image" src="https://user-images.githubusercontent.com/2949987/152677627-79bd9024-d533-4be0-9c57-1c60c340b2b5.png">

Most of the time for `ext4` is spent on formatting the LUN. At the end, both LUN have a different shape in Synology DSM. The `ext4` one looks full, while the `btrfs` one not.

It that an expected behavior?

**Comment by vaskozl** [comment]
I've just tested this with the v1.0.1 and the latest upstream CSI images and the LUNs created with a storage class featuring `fsType: btrfs` do appear full to me.

I haven't noticed ext4 volumes taking any significant time to provision myself.

Any reason to chose ext4 vs btrfs for the fsType, given that the LUNs are already on top of btrfs?

**Comment by chihyuwu** [comment]
Hi @jeanfabrice,
This is an expected behavior **depends on the given filesystem** discards blocks (TRIM) at mkfs time or not. If the filesystem performs whole device TRIM operation, the LUN will look full.

The synology-csi nodeserver runs `mkfs.<fsType>` without any extended-options, so its behavior depends on the default settings of the filesystem.
- For `ext4`, it will attempt to discard blocks at mkfs time by default, see [mkfs.ext4](https://linux.die.net/man/8/mkfs.ext4). An extended-option `nodiscard` can be added to skip discarding blocks.
- For `btrfs`, in my tests, the behaviors seem different according to `btrfs` version, see [btrfs changlog](https://btrfs.wiki.kernel.org/index.php/Changelog#btrfs-progs_v5.15_.28Nov_2021.29).
    - v4.15 - **not** discard blocks
    - v5.15 - discard blocks

I did `mkfs` manually for my thin LUNs and got results:

![20220224_004_SSD](https://user-images.githubusercontent.com/89761945/155503683-343f7ee5-cb02-433c-b74e-9cb75cfb323a.png)

**Comment by vaskozl** [comment]
Thanks for investigating. Is there an advantage to using `discard` for LUNs? I would expect not, if so should we not expose setting nodiscard?

**Comment by jeanfabrice** [comment]
Thanks for the informations @chihyuwu!
I'll stick to `ext4` for now, not sure how useful `btrfs` can be.

**Comment by travisghansen** [comment]
I ran into this developing my (synology compatible) driver here: https://github.com/democratic-csi/democratic-csi

In my case I allow tweaking the formatting parameters by the operator so this particular issue can be avoided.

https://github.com/democratic-csi/democratic-csi/blob/master/examples/node-common.yaml

**Comment by rtim75** [comment]
I'm seeing the same behavior with k3s and 
- ext4
- btrfs
- xfs

all fstypes take quite some time to provision and appear full in the end. I'm using hdds so it takes 5 minutes to provision 30GB

<img width="1125" alt="image" src="https://user-images.githubusercontent.com/29075662/232133824-b1197260-79c1-4912-86cb-dff1d588cba9.png">

**Comment by rtim75** [comment]
by upgrading mount-utils to 0.26.4 I was able to use `FormatAndMountSensitiveWithFormatOptions` and pass `-E nodiscard` options to it, which made plugin create and mount PV a lot faster.
<img width="1110" alt="image" src="https://user-images.githubusercontent.com/29075662/232151195-25a87c73-7e91-4a6f-9420-80450e6990d9.png">

```
I0414 20:37:18.136319       1 mount_linux.go:528] Disk "/dev/disk/by-path/ip----:3260-iscsi-iqn.2000-01.com.synology:---.pvc-b4124636-bff0-4483-8960-01b26669b1f3-lun-1" appears to be unformatted, attempting to format as type: "ext4" with options: [-E nodiscard -F -m0 /dev/disk/by-path/ip----:3260-iscsi-iqn.2000-01.com.synology:---.pvc-b4124636-bff0-4483-8960-01b26669b1f3-lun-1]
I0414 20:37:18.300355       1 mount_linux.go:538] Disk successfully formatted (mkfs): ext4 - /dev/disk/by-path/ip----:3260-iscsi-iqn.2000-01.com.synology:---.pvc-b4124636-bff0-4483-8960-01b26669b1f3-lun-1 /var/lib/kubelet/plugins/kubernetes.io/csi/csi.san.synology.com/6fb05ccd8dba20b798c86f204b1e37d42979beb800a1dc891ff873e0e74500a4/globalmount
```

**Comment by aaron-trout** [comment]
This does not seem to be working for me. I'm using the `synology/synology-csi:v1.1.2` image but the mount options do not appear to be passed to mkfs:

StorageClass:

```yaml
allowVolumeExpansion: true
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  creationTimestamp: "2023-10-07T11:17:24Z"
  name: synology-ext4
  resourceVersion: "720896"
  uid: c1b030bc-0aaa-48aa-bcb8-4005f778f99b
parameters:
  dsm: nas2.lan.trouter.me
  formatOptions: -E nodiscard
  fsType: ext4
  location: /volume1
  protocol: iscsi
provisioner: csi.san.synology.com
reclaimPolicy: Delete
volumeBindingMode: Immediate
```

`ps aux | grep mkfs` on the node while the pod is in `ContainerCreating`:

```
root@k8s-1:~# ps aux | grep mkfs
root     1031787  0.8  0.0   1748  1352 ?        D    11:37   0:00 mkfs.ext4 -F -m0 /dev/disk/by-path/ip-192.168.0.235:3260-iscsi-iqn.2000-01.com.synology:nas2.pvc-ad65563a-02d3-4f51-bd56-f89a678a6fba-lun-1
root     1031961  0.0  0.0   6476  2168 pts/1    S+   11:37   0:00 grep --color=auto mkfs
```

**Comment by aaron-trout** [comment]
Oh, it's just because the PR to introduce `formatOptions` was merged _after_ the 1.1.2 release 😅

@chihyuwu would you be able to mint a v1.1.3? :D Thanks!
